bin/cron/hoc_student_name_cleanup

```diff
@@ -0,0 +1,196 @@
+  start_id: nil,
+  end_id: nil,
+  batch_limit: 100,
+  sleep_msecs: 100,
+  dry_run: false,
+  log_sql: false,
+}
+OptionParser.new do |opts|
+  opts.banner = <<~BANNER
+    Usage: #{File.basename(__FILE__)} [options]
+
+    This script deletes names stored in the pegasus.hoc_activity table after the
+    configured retention period.
+
+    Options:
+  BANNER
+
+  opts.on('--retention=90',
+    Integer,
+    'Retention period in days. Defaults to 90 days if not specified.'
+  ) do |retention_days|
+    options[:retention_days] = retention_days
+  end
+
+  opts.on('--end-id=1000000',
+    Integer,
+    'Id of last entry to clean up (exclusive). If both this and rentention_days ' \
+    'are specified, this value takes precedence.'
+  ) do |end_id|
+    options[:end_id] = end_id
+  end
+
+  opts.on('--start-id=1',
+    Integer,
+    'Id of first entry to clean up (inclusive). If not specified, the script will ' \
+    'determine where to start based on where the last run left off.'
+  ) do |start_id|
+    options[:start_id] = start_id
+  end
+
+  opts.on('--batch-limit=100',
+    Integer,
+    'Maximum number of rows to return in each batch. Defaults to 100 if not specified.'
+  ) do |batch_limit|
+    options[:batch_limit] = batch_limit
+  end
+
+  opts.on('--sleep=100',
+    Integer,
+    'Amount of time to sleep in milliseconds to sleep between batches. Defaults to ' \
+    '100 msecs if not specified.'
+  ) do |sleep_msecs|
+    options[:sleep_msecs] = sleep_msecs
+  end
+
+  opts.on('--dry-run',
+    'Enables read-only mode where no changes are written to the database'
+  ) do |dry_run|
+    options[:dry_run] = dry_run
+  end
+
+  opts.on('--log-sql',
+    'Logs all sql statements sent to the database'
+  ) do |log_sql|
+    options[:log_sql] = log_sql
+  end
+
+  opts.on('-h', '--help', 'Prints this help message') do
+    puts opts
+    exit
+  end
+end.parse!
+CDO.log.info "Initial parsed options: #{options}"
+
+PEGASUS_DB.loggers << CDO.log if options[:log_sql]
+
+# Get the last row in the db for later use. If there is no last row (i.e. there
+# are no rows in the table), we can exit right away.
+last_row = HOC_ACTIVITY.order(:id).last
+if last_row.nil?
+  CDO.log.info "pegasus.hoc_activity table has no rows, exiting"
+  return
+end
+
+# Calculate the start of the retention window. Rows with a timestamp earlier
+# than this are candidates for cleanup.
+options[:retain_after_date] = Date.today - options[:retention_days]
+
+# If start_id was not specified, read the value from the db
+if options[:start_id].nil?
+  options[:start_id] = Properties.get(START_ID_PROPERTY_KEY) || 0
+end
+
+# If end_id was not specified, determine its value by looking for the first row
+# that is definitely inside the retention window. In the rare case where there
+# are no rows in the retention window, set end_id to 1 past the last row to
+# check all rows
+if options[:end_id].nil?
+  options[:end_id] = HOC_ACTIVITY.select(:id).
```

Comment:
My understanding of `Sequel` is poor, but it appears that this queries a potentially large dataset and then grabs the `id` column of the first row in that dataset. We have an index on `finished_at`, but wondering if we can reduce performance impact on the database and memory consumption on the client ruby script, by using a SQL function like `MIN()` or `MAX()` to get the value instead? We have a high write throughput to this table, so minimizing SELECTs on large datasets at the end of the table, would help reduce the impact.

Reply:
This Sequel statement results in the following query:
<pre>
SELECT `id` FROM `hoc_activity` WHERE (`finished_at` > '2020-11-12') ORDER BY `finished_at`, `id` LIMIT 1
</pre>

Does that help alleviate your concern?

Reply:
Yup! That seems much lower impact on both the client and the server.